### PR TITLE
Head parallel and batch fold blocking for Qwen3-VL MOE

### PR DESCRIPTION
**Pull request by kdulla**
# Overview
This PR adds support for new blocking modes, including head parallel blocking, prefill blocking, decode MOE blocking and batch-fold blocking. Head parallel blocking is introduced for causal models while the other approaches are introuced for Qwen3-VL MOE. Expert Parallelism is enabled via expert_parallel compile option while new flags in qaic_config can be passed for different attention blocking configurations. Additionally, this PR contains updated tests for the new blocking modes introduced. 

# Modes Added
- Head parallel softmax kv blocking:
  - split heads across cores, enabled for both causal models and Qwen3VL-moe
  - Enabled by passing kv_blocking_headpar_split (Default: 0 or 16) in qaic config with kv blocking mode
  - invokes newly created blocked_kv_attention_forward_headpar_offline attention forward
- Batch fold + head parallel kv blocking:
  - fold batch size and head dimensions before applying head parallel dimension, intended for high batch size use cases, enabled for Qwen3-VL-moe
  - Enabled by passing batch_fold: True in qaic config with kv blocking mode
  - invokes newly created blocked_kv_attention_forward_decode_headpar_batch attention forward
  - Alters cache layout to be [1, BS * NH, SL, D]
- Online softmax prefill:
  - Applies online softmax to chunked prefill blocks, intended for long-context prefill efficiency
  - Enabled by passing prefill_blocking_mode: "online", prefill_block_chunks, and prefill_n_rep_chunk in qaic_config
  - Invokes newly created blocked_qkv_attention_forward_prefill_online attention forward

Additionally, pre-existing KV blocking modes have been updated to correctly use ctx_len from compile_time for computing block sizes instead of past_seen-tokens. 

# Scripts
E2E run on Qwen3-VL moe with batch fold and prefill blocking can be ran via:
`python examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py`

Unit tests on Qwen3-VL moe can be ran via:
`pytest tests/transformers/models/image_text_to_text/test_qwen3_vl_moe_disagg_blocked.py`

Unit tests for head parallel softmax on causal models added to tests/transformers/models/causal_lm_models/test_causal_lm_blocking_hqkv.py

### CODE REVIEW

File: examples/disagg_serving/gpt_oss_disagg_mode_with_chunking.py [hunk 1]
```diff
@@ -106,6 +110,9 @@
     for i in range(config.num_hidden_layers):
         inputs[f"past_key.{i}"] = qpc_out[f"past_key.{i}_RetainedState"]
         inputs[f"past_value.{i}"] = qpc_out[f"past_value.{i}_RetainedState"]
+print(f"ttft={(time.time() - loop_start):.3f}")
+
+exit(0)
```
Comment (ochougul) [inline examples/disagg_serving/gpt_oss_disagg_mode_with_chunking.py]: ?

File: examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py [hunk 2]
```diff
@@ -217,40 +281,55 @@
...
+        outputs[f"past_value.{layer_idx}_RetainedState"], (1, BS, 1, 1)
+    ).reshape(1, BS * h, c, d)
 
 st = perf_counter()
 decode_out = lang_decode_session.run(decode_inputs)
 print(f"time for first run of decode with KV as input = {perf_counter() - st} sec\n")
 
-all_outputs.append(np.argmax(decode_out["logits"]))
-pos_id = np.max(decode_inputs["position_ids"], axis=-1, keepdims=True) + 1
+# exit(0)
```
Comment (ochougul) [inline examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py]: remove

File: examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py [hunk 3]
```diff
@@ -217,40 +281,55 @@
...
+    "input_ids": np.argmax(decode_out["logits"], axis=-1),  # [BS, 1]
     "position_ids": pos_id,
 }
 
-for i in range(config.text_config.num_hidden_layers):
-    loop_decode_inputs[f"past_key.{i}"] = decode_out[f"past_key.{i}_RetainedState"]
-    loop_decode_inputs[f"past_value.{i}"] = decode_out[f"past_value.{i}_RetainedState"]
+# for i in range(config.text_config.num_hidden_layers):
+#     loop_decode_inputs[f"past_key.{i}"] = decode_out[f"past_key.{i}_RetainedState"]
+#     loop_decode_inputs[f"past_value.{i}"] = decode_out[f"past_value.{i}_RetainedState"]
```
Comment (ochougul) [inline examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py]: ?

File: examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py [hunk 4]
```diff
@@ -217,40 +281,55 @@
...
     decode_out = lang_decode_session.run(loop_decode_inputs)
-    all_outputs.append(np.argmax(decode_out["logits"]))
+    all_outputs.append(np.argmax(decode_out["logits"][0]))
     pos_id += 1
-    for j in range(config.text_config.num_hidden_layers):
-        loop_decode_inputs[f"past_key.{j}"] = decode_out[f"past_key.{j}_RetainedState"]
-        loop_decode_inputs[f"past_value.{j}"] = decode_out[f"past_value.{j}_RetainedState"]
+    # for j in range(config.text_config.num_hidden_layers):
+    #     loop_decode_inputs[f"past_key.{j}"] = decode_out[f"past_key.{j}_RetainedState"]
+    #     loop_decode_inputs[f"past_value.{j}"] = decode_out[f"past_value.{j}_RetainedState"]
```
Comment (ochougul) [inline examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py]: ?

Comment (kdulla) [reply examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py]: Resolved, these were older experiments being done that were incorrectly pushed. Cleaned up example script code accordingly

File: QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py [hunk 5]
```diff
@@ -35,6 +35,7 @@
     BlockingMode,
     generic_blocked_attention_interface,
     past_key_value_update,
+    prefill_blocked_attention_interface,
```
Comment (ochougul) [inline QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py:38]: why are we introducing new interface ? can't it be handled within, we also did siimlar for latent attention?

File: QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py [hunk 6]
```diff
@@ -155,8 +156,8 @@ def forward(self, hidden: torch.Tensor):
         top_w, top_i = torch.topk(router_logits, self.router.top_k, dim=-1)
         top_w = torch.nn.functional.softmax(top_w, dim=1, dtype=top_w.dtype)
 
-        routing_weights = torch.zeros_like(router_logits)
-        routing_weights.scatter_(1, top_i, top_w)
+        # routing_weights = torch.zeros_like(router_logits)
+        # routing_weights.scatter_(1, top_i, top_w)
```
Comment (ochougul) [inline QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py:160]: remove commented code

File: examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py
Comment (ochougul) [inline examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py:1]: Do we really need 3 example files, all of them can be like CLI options for one example file.

Comment (ochougul) [reply examples/image_text_to_text/models/qwen3_vl_moe/qwen3_vl_disagg_mode.py:1]: And default value is the one that is best performant.

File: QEfficient/blocking/blocked_attention_forwards.py [hunk 7]
```diff
@@ -208,6 +247,777 @@ def blocked_kv_attention_forward(
...
+            start_index=start_index,
+        )
+        causal_mask = (
+            causal_mask.expand(batch_size, num_kv_heads, seq_len, kv_len_block)
+            .reshape(1, BH, seq_len, kv_len_block)
+            .unsqueeze(3)
+            .expand(1, BH, seq_len, num_kv_groups, kv_len_block)
+            .reshape(1, BH, seq_len * num_kv_groups, kv_len_block)
+        )
+        attn_weights_block = attn_weights_block.masked_fill(causal_mask, -3.0e4)
```
Comment (ochougul) [inline QEfficient/blocking/blocked_attention_forwards.py]: Please use `MIN_MASKED_ATTENTION_VALUE` here instead of the hardcoded `-3.0e4`. Same for lines 331, 434, 455, 573, 883 and 885. The rest of this file (including the forwards right above and below the new ones) already uses the constant everywhere.

Apart from consistency, mixing the two values is a correctness problem. The running max is seeded with `MIN_MASKED_ATTENTION_VALUE` i.e. `-inf`, but the masked logits get `-3.0e4`. So for a row where the whole block is masked, the block max ends up at `-3.0e4` instead of `-inf`, `exp(logit - max)` becomes 1 for the masked positions instead of 0, and those ones get accumulated into `split_sum`/`split_out` in the two-stage merge.

Also at line 573 please pick up `mask_dtype` the way `blocked_kv_attention_forward` does rather than `query.dtype`, and keep the accumulator seed and the mask fill the same value within each forward.

If `-3.0e4` was picked deliberately because `-inf` was producing nan/inf on device, let me know - in that case it should be a named constant and the seeds should move to match it, not left as a literal.

File: QEfficient/base/modeling_qeff.py [hunk 8]
```diff
@@ -655,6 +673,7 @@ def get_onnx_path(
                     "moe_prefill_packed_chunk_size": constants.MOE_PREFILL_PACKED_CHUNK_SIZE
                     if moe_prefill_packed_chunk_size is None
                     else moe_prefill_packed_chunk_size,
+                    "moe_prefill_target_seq_len": specializations[0].get("seq_len"),
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: This is not needed. what is this `moe_prefill_target_seq_len` we only want `moe_prefill_packed_chunk_size` and `prefill_seq_len` this new variable is not needed.

File: QEfficient/blocking/attention_blocking.py [hunk 9]
```diff
@@ -37,6 +44,15 @@ class BlockingMode(str, Enum):
...
 
+    @classmethod
+    def resolve(cls, mode: Optional[str | "BlockingMode"]) -> "BlockingMode":
+        if mode is None:
+            return cls.NONE
+        resolved_mode = cls(mode)
+        if resolved_mode == cls.AUTO:
+            return cls.HQKV
+        return resolved_mode
+
```
Comment (ochougul) [inline QEfficient/blocking/attention_blocking.py:66]: Why are we adding this as a classmethod here, this decision should have been done before setting any value to `blockingconfig.mode`

File: QEfficient/blocking/attention_blocking.py [hunk 10]
```diff
@@ -46,6 +62,13 @@ class AttentionBlockingConfig:
     head_block_size: Optional[int] = None
     skip_kv: Optional[bool] = True
     num_batch_blocks: Optional[int] = None
+    kv_blocking_headpar_split: Optional[int] = None
+    batch_fold: Optional[bool] = False
+    prefill_block_chunks: Optional[int] = None
+    prefill_blocking_mode: Optional[str] = None  # "q", "kv", "qkv" or "online"
+    prefill_n_rep_chunk: Optional[int] = None
```
Comment (ochougul) [inline QEfficient/blocking/attention_blocking.py]: Do we really need these options again for `prefill_` can't we reuse existing ones?

File: QEfficient/blocking/attention_blocking.py [hunk 11]
```diff
@@ -63,11 +86,30 @@ def supports_blocked_kv(past_key_value: Optional[Cache]) -> bool:
...
     BlockingMode.KV: blocked_kv_mla_attention_forward,
     BlockingMode.H: blocked_h_mla_attention_forward,
 }
 
+_STRATEGIES_PREFILL: Dict[BlockingMode, Callable] = {
+    "q": blocked_q_attention_forward_prefill,
+    "kv": blocked_kv_attention_forward_prefill_headpar_offline,
+    "qkv": blocked_qkv_attention_forward_prefill_headpar_offline,
+    "online": blocked_qkv_attention_forward_prefill_online,
+}
```
Comment (ochougul) [inline QEfficient/blocking/attention_blocking.py]: Type hint is wrong

File: QEfficient/blocking/attention_blocking.py [hunk 12]
```diff
@@ -150,7 +199,12 @@ def generic_blocked_attention_interface(
                 sliding_window=sliding_window,
             )
 
-    strategy = _STRATEGIES.get(blocking_config.mode)
+    if blocking_config.kv_blocking_headpar_split is not None:
+        strategy = _STRATEGIES_HEADPAR.get(blocking_config.mode)
+    elif blocking_config.batch_fold:
+        strategy = _STRATEGIES_BATCH.get(blocking_config.mode)
+    else:
+        strategy = _STRATEGIES.get(blocking_config.mode)
```
Comment (ochougul) [inline QEfficient/blocking/attention_blocking.py]: Why is this if-else needed. We can have a single universal dict and have a method like `resolve_blocking_mode_using_blocking_config` which return the mode and, anybody wanting to add a new method always updates one single DICT.

File: QEfficient/blocking/blocked_attention_forwards.py (hunk 7 above)
Comment (kdulla) [reply QEfficient/blocking/blocked_attention_forwards.py]: -3.0e4 is needed for accuracy, will make it a new constant

File: QEfficient/blocking/attention_blocking.py (hunk 12 above)
Comment (kdulla) [reply QEfficient/blocking/attention_blocking.py]: Will add additional blocking modes for batch fold and head parallel, and reference everything from one dictionary

File: QEfficient/blocking/blocked_attention_forwards.py [hunk 13]
```diff
@@ -42,14 +44,18 @@ def _normalize_int(value: Optional[torch.Tensor | int]) -> int:
     return int(value) if value is not None else 0
 
 
+def _get_headpar_split(configured_split: int, num_kv_groups: int) -> int:
+    return max(1, int(configured_split if configured_split is not None else num_kv_groups))
+
```
Comment (ochougul) [inline QEfficient/blocking/blocked_attention_forwards.py]: remove it.

File: QEfficient/blocking/blocking_configurator.py [hunk 14]
```diff
@@ -342,38 +364,30 @@ def build_transformer_blocking_config_for_transform(
...
+        "prefill_n_rep_chunk",
+        "kv_block_unroll",
+    ):
+        if qaic_config.get(param) is not None:
+            setattr(blocking_config, param, qaic_config.get(param))
+
+    if qaic_config.get("ctx_len") is None:
+        blocking_config.ctx_len = ctx_len
+
+    if qaic_config.get("kv_blocking_headpar_split") is not None:
```
Comment (kdulla) [inline QEfficient/blocking/blocking_configurator.py]: raise error if indivisible kv_blocking_headpar_split is passed

File: QEfficient/blocking/attention_blocking.py [hunk 15]
```diff
@@ -234,3 +291,56 @@ def generic_blocked_mla_attention_interface(
     )
 
     return attn_output, attn_weights
+
+
+def prefill_blocked_attention_interface(
```
Comment (kdulla) [inline QEfficient/blocking/attention_blocking.py]: roll all attention interfaces, MLA and prefill into the generic interface

**Comment by qraniumcitest** [comment]
CI-Ready

**Comment by ochougul** [comment]
CI-Ready

**Comment by ochougul** [comment]
CI-Ready

**Comment by ochougul** [comment]
CI-Ready

**Comment by ochougul** [comment]
CI-Ready

**Comment by vbaddi** [comment]
CI-Ready

**Comment by tv-karthikeya** [comment]
CI-Ready

**Comment by tv-karthikeya** [comment]
CI-Ready

File: QEfficient/base/modeling_qeff.py [hunk 16]
```diff
@@ -646,6 +646,10 @@ def get_onnx_path(
         retain_full_kv: Optional[bool] = False,
         qaic_config: Optional[dict] = None,
         kv_cache_prefix: Optional[str] = None,
+        expert_parallel: Optional[bool] = None,
+        cores_per_expert: Optional[int] = None,
+        tree_reduce: Optional[bool] = None,
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: This is not needed anymore. as we are passing this via qaic_config. Remove it.

File: QEfficient/base/modeling_qeff.py [hunk 17]
```diff
@@ -646,6 +646,10 @@ def get_onnx_path(
...
+        mdp_num_partitions: Optional[int] = 1,
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: Also, this is passed via `**compiler_options` why are we adding it here?

File: QEfficient/base/modeling_qeff.py [hunk 18]
```diff
@@ -1015,6 +1019,7 @@ def _compile(
         qaic_config: Optional[dict] = None,
         specialization_module_name: Optional[str] = None,
         kv_cache_prefix: Optional[str] = None,
+        expert_parallel: Optional[bool] = None,
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: remove

File: QEfficient/blocking/attention_blocking.py [hunk 19]
```diff
@@ -20,22 +20,109 @@
...
+    blocked_q_attention_forward_prefill,
     blocked_qkv_attention_forward,
+    blocked_qkv_attention_forward_prefill_headpar_offline,
+    blocked_qkv_attention_forward_prefill_online,
 )
 
 
 class BlockingMode(str, Enum):
     NONE = ""
+    AUTO = "auto"
```
Comment (ochougul) [inline QEfficient/blocking/attention_blocking.py]: add a comment explaining -> None means no blocking, Auto means Qefficient decides blocking strategy based on VTCM size limit.

File: QEfficient/transformers/models/modeling_auto.py [hunk 20]
```diff
@@ -1602,6 +1603,22 @@ def export(
...
+        }
+        if getattr(self.model.config, "model_type", None) == "qwen3_vl_moe":
+            _blocking_cfg = self.lang_model.hash_params.get("blocking_kwargs", None)
+            batch_fold = (
+                not prefill_only and _blocking_cfg is not None and _blocking_cfg.mode == BlockingMode.KV_BATCH_FOLD
+            )
+            if batch_fold:
+                # Batch-fold is used for the decode graph; export must trace
+                # the same one-token decode shape used by compile/runtime.
+                onnx_kwargs["prefill_seq_len"] = 1 if seq_len else None
```
Comment (ochougul) [inline QEfficient/transformers/models/modeling_auto.py]: why are we keeping `else None` shouldn't we always pass 1 ?

File: QEfficient/transformers/models/modeling_auto.py [hunk 21]
```diff
@@ -1610,11 +1627,7 @@ def export(
                 comp_ctx_lengths=self.comp_ctx_lengths_decode,
                 **onnx_kwargs,
             )
-            dynamic_axes = self.model.get_onnx_dynamic_axes(
-                kv_offload=True,
-                continuous_batching=self.continuous_batching,
-                comp_ctx_lengths=self.comp_ctx_lengths_decode,
-            )
+            dynamic_axes = self.model.get_onnx_dynamic_axes(**dynamic_axes_kwargs)
```
Comment (ochougul) [inline QEfficient/transformers/models/modeling_auto.py]: Don't depend on try/except we should know which path to take based  on if CB was enabled or not.

**Comment by ochougul** [comment]
CI-Ready

**Review by ochougul — APPROVED** [review]

**Comment by tv-karthikeya** [comment]
CI-Ready
